sentry-android-core/src/main/java/io/sentry/android/core/AndroidTraceTransactionListener.java

```diff
@@ -0,0 +1,150 @@
+package io.sentry.android.core;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+import android.os.Build;
+import android.os.Debug;
+import io.sentry.HubAdapter;
+import io.sentry.IHub;
+import io.sentry.ITransaction;
+import io.sentry.ITransactionListener;
+import io.sentry.SentryEnvelope;
+import io.sentry.SentryLevel;
+import io.sentry.SentryOptions;
+import io.sentry.protocol.SentryId;
+import io.sentry.util.FileUtils;
+import io.sentry.util.Objects;
+import io.sentry.util.SentryExecutors;
+import java.io.File;
+import java.io.IOException;
+import java.util.UUID;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class AndroidTraceTransactionListener implements ITransactionListener {
+
+  /**
+   * This appears to correspond to the buffer size of the data part of the file, excluding the key
+   * part. Once the buffer is full, new records are ignored, but the resulting trace file will be
+   * valid.
+   *
+   * <p>30 second traces can require a buffer of a few MB. 8MB is the default buffer size for
+   * [Debug.startMethodTracingSampling], but 3 should be enough for most cases. We can adjust this
+   * in the future if we notice that traces are being truncated in some applications.
+   */
+  private static final int BUFFER_SIZE_BYTES = 3_000_000;
+
+  private final IHub hub;
+
+  private @Nullable File traceFile = null;
+  private @Nullable File traceFilesDir = null;
+  private boolean startedMethodTracing = false;
+  private @Nullable ITransaction activeTransaction = null;
+
+  private @NotNull final SentryOptions options;
+
+  public AndroidTraceTransactionListener() {
+    this(HubAdapter.getInstance());
+  }
+
+  public AndroidTraceTransactionListener(IHub hub) {
+    this.hub = Objects.requireNonNull(hub, "hub is required");
+    options = hub.getOptions();
+    final String tracesFilesDirPath = options.getProfilingTracesDirPath();
+    if (tracesFilesDirPath == null || tracesFilesDirPath.isEmpty()) {
+      options
+          .getLogger()
+          .log(SentryLevel.ERROR, "No profiling traces dir path is defined in options.");
+      return;
+    }
+    traceFilesDir = new File(tracesFilesDirPath);
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public synchronized void onTransactionStart(ITransaction transaction) {
+
+    // Debug.startMethodTracingSampling() is only available since Lollipop
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) return;
+
+    // Let's be sure to end any running trace
+    if (activeTransaction != null) onTransactionEnd(activeTransaction);
```

Comment:
I'd say we should use a boolean to control whether profiling is running or not, if a transaction is already running the profiler, the other ones don't need to, since the profiler is a static method from all threads and not from a specific scope.

Reply:
Just thought about a possible issue: what if the second transaction (that is ignored by the profiler) ends before the first one?
Using a boolean I cannot distinguish the transaction that is finishing the profiling.
So, i think I have 3 options:
1) keep a reference to the transaction: this means that when the transaction that started the profiling finishes, then the profiling finishes, too, even if another transaction is running
2) use a counter, increasing when a transaction starts and decreasing when it finishes: this means that the profiling is stopped when there are no active transactions
3) ignore it and use a boolean: whenever a transaction finishes, the profiling finishes, too

I would go with 1 or 2, depending on whether we want the profiling to cover all running transactions or only the starting one.
What do you think @marandaneto  ?

Reply:
When transaction A starts the profiler, the profiler should be stopped once transaction A is finished.
Either a reference or an identifier is needed, indeed.